From 7a43fc3ed6e307ea7cf987567bada011d58840f9 Mon Sep 17 00:00:00 2001
From: AdarshGrewal <adarshgrewal@gmail.com>
Date: Sat, 10 Aug 2024 14:08:50 +0530
Subject: [PATCH] DNM: SystemUI: Add xiaomi fingerprintextension and
 touchfeature calls

 * before you bully me for this this only till i get fp aidl done

Change-Id: I7e14868f03c7417a413388b0ecb3ca8430165aed
---
 packages/SystemUI/Android.bp                  |  2 +
 .../systemui/biometrics/UdfpsController.java  | 54 +++++++++++++++++++
 services/core/Android.bp                      |  2 +
 3 files changed, 58 insertions(+)

diff --git a/packages/SystemUI/Android.bp b/packages/SystemUI/Android.bp
index 3a7275a877a7..d4cd03f421a4 100644
--- a/packages/SystemUI/Android.bp
+++ b/packages/SystemUI/Android.bp
@@ -590,6 +590,8 @@ android_library {
         "androidx.lifecycle_lifecycle-viewmodel-compose",
         "bcsmartspace-google",
         "vendor.lineage.powershare-V1.0-java",
+        "vendor.xiaomi.hardware.fingerprintextension-V1-java",
+        "vendor.xiaomi.hw.touchfeature-V1-java"
     ],
     libs: [
         "keepanno-annotations",
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
index a393574168dd..e2cb7b394843 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsController.java
@@ -128,6 +128,9 @@
 
 import javax.inject.Inject;
 
+import vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint;
+import vendor.xiaomi.hw.touchfeature.ITouchFeature;
+
 /**
  * Shows and hides the under-display fingerprint sensor (UDFPS) overlay, handles UDFPS touch events,
  * and toggles the UDFPS display mode.
@@ -191,6 +194,9 @@ public class UdfpsController implements DozeReceiver, Dumpable {
     private final boolean mIgnoreRefreshRate;
     private final KeyguardTransitionInteractor mKeyguardTransitionInteractor;
 
+    private static ITouchFeature xaiomiTouchFeatureAidl = null;
+    private static IXiaomiFingerprint xaiomiFingerprintExtensionAidl = null;
+
     // Currently the UdfpsController supports a single UDFPS sensor. If devices have multiple
     // sensors, this, in addition to a lot of the code here, will be updated.
     @VisibleForTesting @NonNull FingerprintSensorPropertiesInternal mSensorProps;
@@ -262,6 +268,50 @@ public void onScreenTurnedOff() {
         }
     };
 
+    private static void xaiomiTouchFeature(int arg) {
+        try {
+            if (xaiomiTouchFeatureAidl == null) {
+                var name = "default";
+                var fqName = vendor.xiaomi.hw.touchfeature.ITouchFeature.DESCRIPTOR + "/" + name;
+                var b = android.os.Binder.allowBlocking(android.os.ServiceManager.waitForDeclaredService(fqName));
+                xaiomiTouchFeatureAidl = vendor.xiaomi.hw.touchfeature.ITouchFeature.Stub.asInterface(b);
+
+                // Link to death
+                b.linkToDeath(() -> {
+                    android.util.Log.w("FP-HAX", "TouchFeature binder died. Reconnecting...");
+                    xaiomiTouchFeatureAidl = null;
+                }, 0);
+
+                android.util.Log.d("FP-HAX", "Binded TouchFeature");
+            }
+            xaiomiTouchFeatureAidl.setTouchMode(0, 10, arg);
+        } catch(Throwable t) {
+            android.util.Log.e("FP-HAX", "TouchFeature", t);
+        }
+    }
+
+    private static void xiaomiFingerprintExtension(int arg) {
+        try {
+            if (xaiomiFingerprintExtensionAidl == null) {
+                var name = "default";
+                var fqName = vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint.DESCRIPTOR + "/" + name;
+                var b = android.os.Binder.allowBlocking(android.os.ServiceManager.waitForDeclaredService(fqName));
+                xaiomiFingerprintExtensionAidl = vendor.xiaomi.hardware.fingerprintextension.IXiaomiFingerprint.Stub.asInterface(b);
+
+                // Link to death
+                b.linkToDeath(() -> {
+                    android.util.Log.w("FP-HAX", "FingerprintExtension binder died. Reconnecting...");
+                    xaiomiFingerprintExtensionAidl = null;
+                }, 0);
+
+                android.util.Log.d("FP-HAX", "Binded FingerprintExtension");
+            }
+            xaiomiFingerprintExtensionAidl.extCmd(4, arg);
+        } catch(Throwable t) {
+            android.util.Log.e("FP-HAX", "FingerprintExtension", t);
+        }
+    }
+
     @Override
     public void dump(@NonNull PrintWriter pw, @NonNull String[] args) {
         final int touchConfigId = mContext.getResources().getInteger(
@@ -1137,6 +1187,8 @@ private void onFingerDown(
                 }
             }
         }
+        xaiomiTouchFeature(1);
+        xiaomiFingerprintExtension(1);
 
         if (isOptical()) {
             for (Callback cb : mCallbacks) {
@@ -1191,6 +1243,8 @@ private void onFingerUp(
         mOnFingerDown = false;
         unconfigureDisplay(view);
         cancelAodSendFingerUpAction();
+        xaiomiTouchFeature(0);
+        xiaomiFingerprintExtension(0);
         if (mPowerManagerInternal != null) {
             mPowerManagerInternal.setPowerMode(PowerManagerInternal.MODE_LAUNCH, false);
         }
diff --git a/services/core/Android.bp b/services/core/Android.bp
index 57afb1302313..ff485c916054 100644
--- a/services/core/Android.bp
+++ b/services/core/Android.bp
@@ -240,6 +240,8 @@ java_library_static {
         "aconfig_new_storage_flags_lib",
         "powerstats_flags_lib",
         "locksettings_flags_lib",
+        "vendor.xiaomi.hardware.fingerprintextension-V1-java",
+        "vendor.xiaomi.hw.touchfeature-V1-java",
     ],
     javac_shard_size: 50,
     javacflags: [
